test/functional/p2p_blocksonly.py

```diff
@@ -57,6 +57,30 @@ def run_test(self):
             self.nodes[0].p2p.wait_for_tx(txid)
             assert_equal(self.nodes[0].getmempoolinfo()['size'], 1)
 
+        self.log.info('Check that txs from whitelisted peers are not rejected and relayed to others')
+        self.log.info("Restarting node 0 with whitelist permission and blocksonly")
+        self.restart_node(0, ["-persistmempool=0", "-whitelist=127.0.0.1", "-whitelistforcerelay", "-blocksonly"])
+        assert_equal(self.nodes[0].getrawmempool(),[])
+        first_peer = self.nodes[0].add_p2p_connection(P2PInterface())
+        second_peer = self.nodes[0].add_p2p_connection(P2PInterface())
+        peer_1_info = self.nodes[0].getpeerinfo()[0]
+        assert_equal(peer_1_info['whitelisted'], True)
+        assert_equal(peer_1_info['permissions'], ['noban', 'forcerelay', 'relay', 'mempool'])
+        peer_2_info = self.nodes[0].getpeerinfo()[1]
+        assert_equal(peer_2_info['whitelisted'], True)
+        assert_equal(peer_2_info['permissions'], ['noban', 'forcerelay', 'relay', 'mempool'])
+        assert_equal(self.nodes[0].testmempoolaccept([sigtx])[0]['allowed'], True)
+        txid = self.nodes[0].testmempoolaccept([sigtx])[0]['txid']
+        first_peer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
+
+        self.log.info('Check that the whitelisted peer is still connected after sending the transaction')
+        assert_equal(first_peer.is_connected, True)
+
+        self.log.info('Check that the tx from whitelisted first_peer is relayed to others (ie.second_peer)')
+        with self.nodes[0].assert_debug_log(["received getdata"]):
```

Comment:
```suggestion
        with self.nodes[0].assert_debug_log(["received getdata", "Force relaying tx {} from whitelisted peer={}".format(txid, 0)]):
```

You can assert that the if-branch is hit by reading the log statement that is written to the debug log when this if-branch is executed.

Reply:
I am still not sure why I am getting error trying to assert that this if branch is hit. 

From ` else if (permission == "forcerelay") NetPermissions::AddFlag(flags, PF_FORCERELAY);` I konw that we definitely have PF_FORCERELAY flag. And this if branch should be hit. However, when I tried to assert this, I got this following error:

AssertionError: [node 0] Expected messages "['received getdata', 'Force relaying tx af34fc5ff9ea8babbd4083fbb79ffd2ad5aff1d6def803c07ca5aeed880bd60f from whitelisted peer=0']" does not partially match log:

 - 2020-04-29T05:33:05.256801Z [msghand] Enqueuing TransactionAddedToMempool: txid=af34fc5ff9ea8babbd4083fbb79ffd2ad5aff1d6def803c07ca5aeed880bd60f wtxid=af34fc5ff9ea8babbd4083fbb79ffd2ad5aff1d6def803c07ca5aeed880bd60f
 - 2020-04-29T05:33:05.257550Z [msghand] Checking mempool with 1 transactions and 1 inputs
 - 2020-04-29T05:33:05.257628Z [msghand] AcceptToMemoryPool: peer=0: accepted af34fc5ff9ea8babbd4083fbb79ffd2ad5aff1d6def803c07ca5aeed880bd60f (poolsz 1 txn, 1 kB)
 - 2020-04-29T05:33:05.257685Z [msghand] sending inv (37 bytes) peer=1
 - 2020-04-29T05:33:05.257745Z [scheduler] TransactionAddedToMempool: txid=af34fc5ff9ea8babbd4083fbb79ffd2ad5aff1d6def803c07ca5aeed880bd60f wtxid=af34fc5ff9ea8babbd4083fbb79ffd2ad5aff1d6def803c07ca5aeed880bd60f
 - 2020-04-29T05:33:05.257955Z [scheduler] [default wallet] AddToWallet af34fc5ff9ea8babbd4083fbb79ffd2ad5aff1d6def803c07ca5aeed880bd60f
 - 2020-04-29T05:33:05.258554Z [msghand] received: getdata (37 bytes) peer=1
 - 2020-04

Reply:
And the other place where we tested whether this if-branch is hit is in p2p_permissions.py: and the major difference between the two setups is this: 
`p2p_rebroadcast_wallet = self.nodes[1].add_p2p_connection(P2PDataStore())` 

Also, this test was able to assert `Force relaying ...` beacuse this broadcast happened between two full-nodes, not between full node and mini-node (like the case here). However, I did try substituing one full node out with a mini node, and this didn't change anything and test still passed.

Reply:
You will need to send the message (tx) inside the assert debug log context mangager. Otherwise the log is not captured.